File: oolab/src/main/java/agh/ics/oop/model/Vector2d.java
package agh.ics.oop.model;

import java.util.Objects;

public class Vector2d {
    private final int x;
    private final int y;

    public Vector2d(int x, int y) {
        this.x = x;
        this.y = y;
    }

    int getX() {
        return this.x;
    }

    int getY() {
        return this.y;
    }

    public String toString() {
        return String.format("(%d,%d)", this.getX(), this.getY());
    }

    public boolean precedes(Vector2d other) {
        return this.getX() <= other.getX() && this.getY() <= other.getY();
    }

    public boolean follows(Vector2d other) {
        return this.getX() >= other.getX() && this.getY() >= other.getY();
    }

    public Vector2d add(Vector2d other) {
        return new Vector2d(this.getX() + other.getX(), this.getY() + other.getY());
    }

    public Vector2d subtract(Vector2d other) {
        return new Vector2d(this.getX() - other.getX(), this.getY() - other.getY());
    }

    public Vector2d upperRight(Vector2d other) {
        return new Vector2d(Math.max(this.getX(), other.getX()), Math.max(this.getY(), other.getY()));
    }

    public Vector2d lowerLeft(Vector2d other) {
        return new Vector2d(Math.min(this.getX(), other.getX()), Math.min(this.getY(), other.getY()));
    }

    public Vector2d opposite() {
        return new Vector2d(-this.getX(), -this.getY());
    }

    public boolean equals(Object other) {
        if (this == other)
            return true;
        if (!(other instanceof Vector2d that))
            return false;
        return Objects.equals(this.getX(), that.getX()) && Objects.equals(this.getY(), that.getY());
    }

    public int hashCode() {
        return Objects.hash(this.getX(), this.getY());
    }
}

File: OOKotlin/src/main/kotlin/Vector2d.kt
import kotlin.math.max
import kotlin.math.min

data class Vector2d(
    val x: Int, val y: Int
) {
    companion object {
        fun MapDirection.toUnitVector() = when (this) {
            MapDirection.NORTH -> Vector2d(0, 1);
            MapDirection.EAST -> Vector2d(1, 0);
            MapDirection.SOUTH -> Vector2d(0, -1);
            MapDirection.WEST -> Vector2d(-1, 0);
        };
    }

    override fun toString() = "$x, $y"

    operator fun compareTo(other: Vector2d): Int {
        val diffX = x - other.x
        val diffY = x - other.x
        if (diffX > 0 && diffY > 0) return 1
        if (diffX < 0 && diffY < 0) return -1
        return 0
    }

//    fun precedes(other: Vector2d) = this < other;

//    fun follows(other: Vector2d) = this > other;

    operator fun plus(other: Vector2d) = Vector2d(x + other.x, y + other.y);

    operator fun minus(other: Vector2d) = Vector2d(x - other.x, y - other.y);

    fun upperRight(other: Vector2d) =
        Vector2d(max(x, other.x), max(y, other.y));

    fun lowerLeft(other: Vector2d) = Vector2d(min(x, other.x), min(y, other.y));

    fun opposite() = Vector2d(-x, -y);

    fun isLegal(bottomLeftBound: Vector2d, upperRightBound: Vector2d) =
        bottomLeftBound < this && this < upperRightBound


}
